addon/webextension/background/analytics.js

```diff
@@ -8,34 +8,70 @@ this.analytics = (function() {
   let telemetryPrefKnown = false;
   let telemetryPref;
 
+  let events = [];
+  let timings = [];
+  let eventsIdleCallbackHandle, timingsIdleCallbackHandle;
+  const fetchOptions = {
+    method: "POST",
+    mode: "cors",
+    headers: { "content-type": "application/json" },
+    credentials: "include"
+  };
+
+  function flushEvents() {
+    if (events.length === 0) {
+      return;
+    }
+
+    let eventsUrl = `${main.getBackend()}/event`;
+    let deviceId = auth.getDeviceId();
+    let sendTime = Date.now();
+
+    events.forEach(event => {
+      event.queueTime = sendTime - event.eventTime
+      log.info(`sendEvent ${event.event}/${event.action}/${event.label || 'none'} ${JSON.stringify(event.options)}`);
+    });
+
+    let body = JSON.stringify({deviceId, events});
+    let fetchRequest = fetch(eventsUrl, Object.assign({body}, fetchOptions));
+    fetchWatcher(fetchRequest);
+
+    events = [];
+    eventsIdleCallbackHandle = null;
+  }
+
+  function flushTimings() {
+    if (timings.length === 0) {
+      return;
+    }
+
+    let timingsUrl = `${main.getBackend()}/timing`;
+    let deviceId = auth.getDeviceId();
+    let body = JSON.stringify({deviceId, timings});
+    let fetchRequest = fetch(timingsUrl, Object.assign({body}, fetchOptions));
+    fetchWatcher(fetchRequest);
+
+    timings.forEach(t => {
+      log.info(`sendTiming ${t.timingCategory}/${t.timingLabel}/${t.timingVar}: ${t.timingValue}`);
+    });
+
+    timings = [];
+    timingsIdleCallbackHandle = null;
+  }
+
   function sendTiming(timingLabel, timingVar, timingValue) {
     // sendTiming is only called in response to sendEvent, so no need to check
     // the telemetry pref again here.
     let timingCategory = "addon";
-    return new Promise((resolve, reject) => {
-      let url = main.getBackend() + "/timing";
-      let req = new XMLHttpRequest();
-      req.open("POST", url);
-      req.setRequestHeader("content-type", "application/json");
-      req.onload = catcher.watchFunction(() => {
-        if (req.status >= 300) {
-          let exc = new Error("Bad response from POST /timing");
-          exc.status = req.status;
-          exc.statusText = req.statusText;
-          reject(exc);
-        } else {
-          resolve();
-        }
-      });
-      log.info(`sendTiming ${timingCategory}/${timingLabel}/${timingVar}: ${timingValue}`);
-      req.send(JSON.stringify({
-        deviceId: auth.getDeviceId(),
-        timingCategory,
-        timingLabel,
-        timingVar,
-        timingValue
-      }));
+    timings.push({
+      timingCategory,
+      timingLabel,
+      timingVar,
+      timingValue
     });
+    if (!timingsIdleCallbackHandle) {
+      window.requestIdleCallback(flushTimings);
```

Comment:
Reading about [requestIdleCallback](https://developer.mozilla.org/en-US/docs/Web/API/Window/requestIdleCallback), it doesn't really seem like the right thing for what we're doing.

We *want* to batch timings, not just to avoid doing work when the code isn't idle, but also to reduce overall network. There is a substantial HTTP overhead to each request after all.

My reading of requestIdleCallback is that it can be run extremely soon, and doesn't have any real delay. There's a `timeout` parameter, but that's a maximum time, not a minimum time. If we were being fancy we'd actually want to do both `setTimeout` (to establish a minimum) and `requestIdleCallback` (to schedule the task during idle time).